llvm/lib/CodeGen/GlobalISel/CallLowering.cpp

```diff
@@ -790,6 +804,32 @@ bool CallLowering::handleAssignments(ValueHandler &Handler,
       CCValAssign &VA = ArgLocs[j + Idx];
       const ISD::ArgFlagsTy Flags = Args[i].Flags[Part];
 
+      // We found an indirect parameter passing and we are at the first part of
+      // the value being passed. In this case copy the incoming pointer into a
+      // virtual register so later we can load it.
+      if (VA.getLocInfo() == CCValAssign::Indirect && Flags.isSplit()) {
+        IndirectParameterPassingHandled = true;
+        if (Handler.isIncomingArgumentHandler()) {
+          Handler.assignValueToReg(ArgReg, VA.getLocReg(), VA);
+          Handler.assignValueToAddress(Args[i].OrigRegs[Part],
+                                       Args[i].Regs[Part], OrigTy,
+                                       MachinePointerInfo{}, VA);
+        } else {
+          Align StackAlign = DL.getPrefTypeAlign(Args[i].Ty);
+          MachineFrameInfo &MFI = MF.getFrameInfo();
+          int FrameIdx = MFI.CreateStackObject(OrigTy.getScalarSizeInBits(),
+                                               StackAlign, false);
+
+          Register PointerToStackReg =
+              MIRBuilder.buildFrameIndex(PointerTy, FrameIdx).getReg(0);
+          Handler.assignValueToAddress(Args[i].OrigRegs[Part],
+                                       PointerToStackReg, OrigTy,
+                                       MachinePointerInfo{}, VA);
+          Handler.assignValueToReg(PointerToStackReg, VA.getLocReg(), VA);
```

Comment:
Calling two assign* methods for single VA doesn't look right. VA has single location, which is either memory (VA.isMemLoc() returns true) or a register, but not both.
I think you should store the temporary object directly using G_STORE and call single assign* method based on the kind of the location (register/memory), passing it pointer to the temporary object.

Same applies to incoming arguments.

Reply:
So I should build my own `G_STORE` in the `handleAssignments` part. When I first started working on this PR I have done that, but later replaced to the callback. I will rewrite it so it will build the G_STORE and not dispatch it ti the target specific handler.

Reply:
I have changed the loads and stores, so the creation of them is not dispatched to the target specific handler. I have also played around with the alignments a bit.

Reply:
This still does not check that VA is a register location. It can be a memory location if all registers used for passing arguments are used (need a test for this).

Reply:
The inner loop that loop over the parts calls a `break;` in case of indirect passing, so we wont loop over all of the parts.

If all registers are used, then it is not an indirect parameter passing (example: when we have a 64 bit reg on rv32 then it is passed in two args not indirectly. In this case the if is not entered)

Yes I should check if the VA is register allocated.

Also maybe I should move this logic out of the loop, since  if we indirect passing then just the 0th element is "processed".

Reply:
What I mean is that you can have e.g.:
define void @test(i64 %x0, i64 %x1, i64 %x2, ...., i128 %y)
`x` arguments occupy all registers available for argument passing. `y` is still passed indirectly, but the pointer to it is not in a register, but in memory.

Reply:
Oh now I understand. I will add more checks and tests then.

Reply:
I think I solved this. Please could you check it? (I am pretty new to compiler back-ends, so I am not 100% sure.)

Now for this IR:
```llvm
define i64 @fun(i64 %x1, i64 %x2, i64 %x3, i64 %x4, i128 %y ) {
  %2 = trunc i128 %y to i64
  ret i64 %2
}

define i32 @fun_caller( ) {
  %1 = call i64 @fun(i64 1,i64 1, i64 1, i64 1, i128 2)
  %2 = trunc i64 %1 to i32
  ret i32 %2
}
```

We generate this:
Callee:
```llvm
Frame Objects:
  fi#-1: size=4, align=16, fixed, at location [SP]
Function Live Ins: $x10, $x11, $x12, $x13, $x14, $x15, $x16, $x17, %14

bb.1 (%ir-block.0):
  liveins: $v3, $x10, $x11, $x12, $x13, $x14, $x15, $x16, $x17
  %5:_(s32) = COPY $x10
  %6:_(s32) = COPY $x11
  %0:_(s64) = G_MERGE_VALUES %5:_(s32), %6:_(s32)
  %7:_(s32) = COPY $x12
  %8:_(s32) = COPY $x13
  %1:_(s64) = G_MERGE_VALUES %7:_(s32), %8:_(s32)
  %9:_(s32) = COPY $x14
  %10:_(s32) = COPY $x15
  %2:_(s64) = G_MERGE_VALUES %9:_(s32), %10:_(s32)
  %11:_(s32) = COPY $x16
  %12:_(s32) = COPY $x17
  %3:_(s64) = G_MERGE_VALUES %11:_(s32), %12:_(s32)
  %14:_(p0) = G_FRAME_INDEX %fixed-stack.0
  %13:_(p0) = COPY %14:_(p0)
  %4:_(s128) = G_LOAD %13:_(p0) :: (load (s128), align 8)
  %15:_(s64) = G_TRUNC %4:_(s128)
  %16:_(s32), %17:_(s32) = G_UNMERGE_VALUES %15:_(s64)
  $x10 = COPY %16:_(s32)
  $x11 = COPY %17:_(s32)
  PseudoRET implicit $x10, implicit $x11
```

From this the relevant part for us the indirect argument receiving looks like this:
```llvm
  %14:_(p0) = G_FRAME_INDEX %fixed-stack.0
  %13:_(p0) = COPY %14:_(p0)
  %4:_(s128) = G_LOAD %13:_(p0) :: (load (s128), align 8)
  %15:_(s64) = G_TRUNC %4:_(s128)
```

Caller:
```llvm
Frame Objects:
  fi#0: size=128, align=8, at location [SP]

bb.1 (%ir-block.0):
  %1:_(s64) = G_CONSTANT i64 1
  %2:_(s128) = G_CONSTANT i128 2
  ADJCALLSTACKDOWN 4, 0, implicit-def $x2, implicit $x2
  %3:_(s32), %4:_(s32) = G_UNMERGE_VALUES %1:_(s64)
  %5:_(s32), %6:_(s32) = G_UNMERGE_VALUES %1:_(s64)
  %7:_(s32), %8:_(s32) = G_UNMERGE_VALUES %1:_(s64)
  %9:_(s32), %10:_(s32) = G_UNMERGE_VALUES %1:_(s64)
  %11:_(p0) = COPY $x2
  %12:_(s32) = G_CONSTANT i32 0
  %13:_(p0) = G_PTR_ADD %11:_, %12:_(s32)
  %14:_(p0) = G_FRAME_INDEX %stack.0
  G_STORE %2:_(s128), %14:_(p0) :: (store (s128), align 8)
  G_STORE %14:_(p0), %13:_(p0) :: (store (p0), align 8)
  $x10 = COPY %3:_(s32)
  $x11 = COPY %4:_(s32)
  $x12 = COPY %5:_(s32)
  $x13 = COPY %6:_(s32)
  $x14 = COPY %7:_(s32)
  $x15 = COPY %8:_(s32)
  $x16 = COPY %9:_(s32)
  $x17 = COPY %10:_(s32)
  PseudoCALL target-flags(riscv-call) @fun, <regmask $vlenb $x0 $x1 $x8 $x9 $x18 $x19 $x20 $x21 $x22 $x23 $x24 $x25 $x26 $x27 $x8_x9 $x18_x19 $x20_x21 $x22_x23 $x24_x25 $x26_x27>, implicit-def $x1, implicit $x10, implicit $x11, implicit $x12, implicit $x13, implicit $x14, implicit $x15, implicit $x16, implicit $x17, implicit-def $x10, implicit-def $x11
  ADJCALLSTACKUP 4, 0, implicit-def $x2, implicit $x2
  %15:_(s32) = COPY $x10
  %16:_(s32) = COPY $x11
  %0:_(s64) = G_MERGE_VALUES %15:_(s32), %16:_(s32)
  %17:_(s32) = G_TRUNC %0:_(s64)
  $x10 = COPY %17:_(s32)
  PseudoRET implicit $x10
```

The relevant part for argument passing:
```llvm
  %11:_(p0) = COPY $x2
  %12:_(s32) = G_CONSTANT i32 0
  %13:_(p0) = G_PTR_ADD %11:_, %12:_(s32)
  %14:_(p0) = G_FRAME_INDEX %stack.0
  G_STORE %2:_(s128), %14:_(p0) :: (store (s128), align 8)
  G_STORE %14:_(p0), %13:_(p0) :: (store (p0), align 8)
```

Reply:
The IR looks correct, except for %14 in the live-ins list.

Reply:
I see. I re-run llc on the above provided IR and the %14 is still in the live-ins. Meanwhile looking at the new tests. There is no virtual register for the live-ins. And the tests are passing. I am looking into it.